src/drivers/uavcan_v1/Publishers/Publisher.hpp

```diff
@@ -0,0 +1,97 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file Publisher.hpp
+ *
+ * Defines basic functionality of UAVCAN v1 publisher class
+ *
+ * @author Jacob Crabill <jacob@flyvoly.com>
+ */
+
+#pragma once
+
+#include <px4_platform_common/px4_config.h>
+
+#include <lib/parameters/param.h>
+
+#include <uavcan/_register/Access_1_0.h>
+
+#include "../CanardInterface.hpp"
+#include "../ParamManager.hpp"
+
+class UavcanPublication
+{
+public:
+	UavcanPublication(CanardInstance &ins, UavcanParamManager &pmgr, const char *uavcan_pname) :
+		_canard_instance(ins), _param_manager(pmgr), _uavcan_param(uavcan_pname) { };
```

Comment:
This could be done better from the standpoint of the principle of single responsibility. Consider replacing the parameter name with **subject name** here, such that you would determine the parameter name afterward like `uavcan.pub.SUBJECT_NAME.id`. Eventually it would allow one to add support for type introspection by automatically exposing a read-only register `uavcan.pub.SUBJECT_NAME.type` containing the name of the published type.

Reply:
Done